Vulkan acceleration structure manager

cmannett85-arm opened the pull request:
Manages the capturing, serialising, and replay of acceleration structures for Vulkan.  It works in a similar way to how device memory is handled:
* A temporary host-accessible buffer is created for each AS
* The AS is serialised (in the Vulkan AS sense of the word) into it
* The buffer handle is stored in the initial contents and downloaded when appropriate
* Replay is handled similarly but in reverse

Note this is not currently wired up to the init state or descriptor functions yet - so doesn't do anything.  The complete work (coming in subsequent PRs) has been tested on a Vivo X90 Pro running Android 13 and on an RTX 3090 FE on Windows 11.

The serialiser version has been bumped for backwards compatibility with 'RenderDoc for Arm GPUs 2024.0'.

Normally this kind of functionality would be done directly in `WrappedVulkan`, I moved it out into its own type in order to ease merges from upstream but ended up keeping it as it's nicer to work with.  However the downside to that is some previously private `WrappedVulkan` methods have had to be made public - I'm happy to move everything inside `WrappedVulkan` if desired.

oscarbg commented:
Hi @cmannett85-arm [cmannett85-arm](https://github.com/cmannett85-arm),
sorry to join here but just was going to open an issue requesting here for ARM engineers to submit here PR changes from "RenderDoc for Arm GPUs" that is based on upstream RenderDoc 1.31, and has capture support for:
VK_KHR_ray_query and VK_KHR_acceleration_structure for device-side acceleration structure builds.

so basically enabling:
1)VK_KHR_ray_query 
2)VK_KHR_acceleration_structure
seems this pr is part of 2)..

also sorry I know that is not a support forum for "RenderDoc for Arm GPUs 2024.0" but I have been testing two ray_query VK PC apps on "RenderDoc for Arm GPUs 2024.0" and they fail (either crash o no capture shown) on NV 4070 with latest NV drivers 551.81..
note I checked "RenderDoc for Arm GPUs 2024.0" works correctly on my PC capturing non ray_query enabled apps like vulkan samples afbc sample (https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/performance/afbc) so the errors are on ray query apps only..

is this usage, i.e. capturing traces of PC VK ray query enabled apps, expected to work with "RenderDoc for Arm GPUs 2024.0"?
two examples failing:
1) testing on Q2RTX 1.7 with "ray queries" mode crashes..
see crash file:
[Q2RTX_CrashReport00.txt](https://github.com/baldurk/renderdoc/files/14549598/Q2RTX_CrashReport00.txt)

2) testing on official vulkan samples ray_queries sample fails (https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/ray_queries):
settings for launcing:
![Captura de pantalla 2024-03-10 101214](https://github.com/baldurk/renderdoc/assets/669305/7020ee26-ca5d-40b7-897a-1a5d036f20ce)
stuck screen upon screen capture and then closes app and no capture snapshot taken:
![Captura de pantalla 2024-03-10 101223](https://github.com/baldurk/renderdoc/assets/669305/4116883f-d8f0-440d-ab65-f54fd716523d)

cmannett85-arm commented:
@oscarbg this is not the place to post bug reports, firstly it's a PR, secondly this is for RenderDoc not 'RenderDoc for Arm GPUs'.  Please submit this report to either developer@arm.com or the [support forum](https://community.arm.com/support-forums/f/graphics-gaming-and-vr-forum).

Although I can tell you now, that RQ captures using RenderDoc for Arm GPUs on an NV card won't work because our RQ capturing requires host-accesible memory support for all memory.  Which is fine for phones which share their RAM between the CPU and GPU - not fine for discrete cards.

This PR (and a couple more to follow) is for porting RQ support to upstream and _will_ include discrete card support.

oscarbg commented:
sorry @cmannett85-arm.. but also thanks for sharing the current limitations and upcoming desktop GPU support in official renderdoc..

baldurk commented on renderdoc/driver/vulkan/renderdoc_vulkan.vcxproj:
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
```
It looks like a BOM marker has been added here, please revert that.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -2460,10 +2492,13 @@ bool WrappedVulkan::EndFrameCapture(DeviceOwnedWindow devWnd)
-    vkr = vt->MapMemory(Unwrap(device), Unwrap(readbackMem.mem), readbackMem.offs, readbackMem.size,
-                        0, (void **)&pData);
+    vkr = vt->MapMemory(Unwrap(device), Unwrap(readbackMem.mem), readbackMem.offs, alignedSize, 0,
+                        (void **)&pData);
```
What's the reason for this change? The `readbackMem` memory size returned from `AllocateMemoryForResource` above should already be aligned to the non-coherent atom size.

cmannett85-arm replied:
The validation layer was reporting a misalignment until I set this.  The size _allocated_ is mem aligned (vk_memory.cpp:172) but the size returned from `AllocateMemoryForResource(..)` is the requested size (vk_memory.cpp:394) which isn't necessarily aligned.  In this case `GetByteSize(..)` may be 'naturally' aligned due to what it's doing, but it isn't explicitly aligned.  At least that was my understanding of the code!

baldurk replied:
Oh I see what you mean, yes you're right. I thought I had fixed cases like this previously but clearly not - I guess non-coherent memory is rare so wouldn't come up often.

baldurk commented on renderdoc/driver/vulkan/vk_common.cpp:
```diff
@@ -362,6 +362,10 @@ bool VkInitParams::IsSupportedVersion(uint64_t ver)
   if(ver == CurrentVersion)
     return true;
 
+  // 0x15 -> 0x16 - courtesy bump for RenderDoc for Arm GPUs 2024.0 compatibility with upstream
```
I think this should just be noted as "added support for acceleration structures". The fact that we're bumping to allow room for compatibility is much of a muchness, but that will make much more sense to most people rather than this which few have the context for.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,548 @@
+}
+
+VulkanAccelerationStructureManager::VulkanAccelerationStructureManager(WrappedVulkan *parent)
+    : m_Parent{parent}
```
Please use a normal constructor here, `m_Parent(parent)`.

Also by entirely arbitrary convention the wrapped vulkan instance usually exists as `m_pDriver` or `m_Driver` rather than parent which implies a more explicit parent-child relationship than exists here really.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,548 @@
+      m_Parent->CheckVkResult(vkr);
+
+      // invalidate the cpu cache for this memory range to avoid reading stale data
+      const VkMappedMemoryRange range{
```
Please use normal variable initialisation, this should be `const VkMappedMemoryRange range = {`.

(this isn't the only case I'm picking one at random to comment on).

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,548 @@
+                                                   const VkInitialContents *initial,
+                                                   CaptureState state)
+{
+  if(ser.VersionLess(version_2024_0))
```
This check shouldn't exist upstream, it should only be on your fork for handling old serialisation.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,548 @@
+
+  SERIALISE_ELEMENT_LOCAL(tag, initial->tag);
+  rdcliteral name =
+      tag == VkInitialContents::SerialisedBottomLevelAccelerationStructure ? "BLAS"_lit : "TLAS"_lit;
```
This serialised tag does not seem to be used as part of the serialisation so I think it should be dropped. The distinction between TLAS and BLAS initial contents seems to only be for ensuring the apply order, unless I'm mistaken. I think it would be better to have this be a bool flag in the VkInitialContents rather than two fundamentally different types (the initial contents types should generally be 1:1 with the resource types). It doesn't have to be serialised, it can be determined by checking if the handle count is non-zero.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,548 @@
+    m_Parent->CheckVkResult(vkr);
+
+    uploadMemory = m_Parent->AllocateMemoryForResource(uploadBuf, MemoryScope::InitialContents,
+                                                       MemoryType::Upload);
```
`AllocateMemoryForResource` does not currently set the right bits for BDA buffers to be bound to it.

cmannett85-arm replied:
It will be in a subsequent PR (https://github.com/cmannett85-arm/renderdoc/blob/as_manager_2/renderdoc/driver/vulkan/vk_memory.cpp#L331).  This could be too heavy handed, but we'll cross that bridge when I PR it.

baldurk reviewed (changes requested):
There are some misunderstandings and a bit of a wrong direction here. I'm not 100% clear since the code that uses this isn't all there yet such as the call to ProcessDescriptorSetSlots, but this doesn't seem like the right organisation.

The way you are using forced references is not correct - you seem to be using them the way that normal frame references work by enumerating the list of referenced resources during the capture and then clearing them when the capture is finished. The point of forced references is to sidestep the need to enumerate resources for marking them referenced when it's not feasible or worthwhile - currently that's for BDA but I would suggest doing the same with BLASs.

As a result the two-stage approach of serialising TLASs and then determining which BLASs they reference and serialise those is more complicated than it needs to be.

You should rejig this to be simpler: TLASs referenced directly from descriptor sets are marked as frame referenced in the same way as any other descriptor-referenced resource. Their initial states are prepared and serialised as-is. BLASs are forced referenced the same way that BDA buffers are and prepared and serialised as-is. With that way there's no need for any synchronisation or co-ordination between them, and there should be no need for any of the tracking and lookups you have currently. Because these exist as actual objects and because we're serialising their contents directly for now we don't need any extra tracking.

This should make for a simpler change that fits naturally into the existing way of serialising resource initial states rather than creating a new side-band way to serialise AS initial states. The buffer management around prepare, serialise, and apply looks OK to me, but I'll give that code a more thorough check after the above refactors.

cmannett85-arm commented:
I'm still suffering from an intermittent `VK_ERROR_DEVICE_LOST` failure, but hopefully another pair of eyes can spot where I'm going wrong.  It occurs during the [serialising of the AS to the readback mem](https://github.com/baldurk/renderdoc/pull/3271/files#diff-0fc27a78a09357936e822ace8f29fee4f432a61c7a42113f780104ccbf07911cR90) during initial state preparation, but not at first, at some (seemingly) arbitrary number of copies in.  The number depending upon the complexity of the scene.

Also the `VK_ERROR_DEVICE_LOST` state is first noticed much later in the APK (usually from a `vkGetFenceStatus`), but I can see from Mali's dmesg output that it occurs in `WrappedVulkan::SubmitCmds(..)` following the `vkCmdCopyAccelerationStructureToMemoryKHR(..)` call.  There's nothing else in the command buffer at this point because it's flushed before and after.

I can't get it to fail in the same way on my nVidia RTX 3090FE, but it does fail on capture load via a segfault in the driver when calling `vkCmdBuildAccelerationStructuresKHR(..)`.  There's nothing obvious in the vars at the crash site, but it is preceded by this in the log:
```
RDOC 042940: [10:16:53]   vk_queue_funcs.cpp(1320) - Debug   - Previous queue execution was on queue ResourceId::1000000000000000543, now executing ResourceId::1000000000000000544, syncing GPU
```
I've no idea if these two issues are related.  My hunch is that I need some Vulkan sync in some places, but I can't see why and there's no obvious precedent in the existing code.

baldurk commented on renderdoc/driver/vulkan/renderdoc_vulkan.vcxproj:
```diff
@@ -1,4 +1,3 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
```
You've removed the `<?xml>` tag entirely now :).

cmannett85-arm replied:
Ha, can you tell I'm not developing on Windows!?

baldurk commented on renderdoc/driver/vulkan/vk_common.cpp:
```diff
@@ -1172,7 +1176,8 @@ FrameRefType GetRefType(DescriptorSlotType descType)
     case DescriptorSlotType::StorageBuffer:
-    case DescriptorSlotType::StorageBufferDynamic: return eFrameRef_ReadBeforeWrite;
+    case DescriptorSlotType::StorageBufferDynamic:
+    case DescriptorSlotType::AccelerationStructure: return eFrameRef_ReadBeforeWrite;
```
As far as I know Acceleration structures are read-only? This will be important for optimisation later, as resources which are read-only throughout an entire frame can have their initial contents applied only once rather than every time since we know they won't mutate after the first application.

cmannett85-arm replied:
They can be updated midframe with the same resource handle when the build mode is `VK_BUILD_ACCELERATION_STRUCTURE_MODE_UPDATE_KHR`, [with caveats](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap37.html#VkAccelerationStructureBuildGeometryInfoKHR).

baldurk replied:
Right ASs can be changed during a capture, as can uniform buffers or images, but those modifications would be tracked via frame references in those specific functions (the build or copy functions for ASs). The _descriptor_ access is always read-only which is what this function is trying to determine.

cmannett85-arm replied:
Ah I understand now, yes AFAIK you can't modify the AS from a descriptor.  I'll update the code.

baldurk commented on renderdoc/driver/vulkan/vk_common.cpp:
```diff
@@ -1207,6 +1212,13 @@ void DescriptorSetSlot::SetTexelBuffer(VkDescriptorType writeType, ResourceId id
+                                                 VkAccelerationStructureKHR accelerationStructure)
+{
+  type = convert(writeType);
+  resource = GetResID(accelerationStructure);
```
As a minor thing it would be good to add 'AS' to the comment on resource to let people know what is stored there. It's fairly self-evident but good to be explicit.

baldurk commented on renderdoc/driver/vulkan/vk_core.h:
```diff
@@ -1352,6 +1350,12 @@ class WrappedVulkan : public IFrameCapturer
     return NULL;
   }
 
+  void AddForcedReference(VkResourceRecord *record)
```
It doesn't matter too much, but just for the sake of removing unnecessary changes I don't think this function needs to be public anymore.

baldurk commented on renderdoc/driver/vulkan/vk_memory.cpp:
```diff
@@ -319,9 +319,16 @@ MemoryAllocation WrappedVulkan::AllocateMemoryForResource(bool buffer, VkMemoryR
 
+    // if ray tracing acceleration structures are in use, then allocate memory with buffer device
+    // address support enabled
+    VkMemoryAllocateFlagsInfo flagsInfo{
```
Please use normal initialisation without the omitted =

```suggestion
    VkMemoryAllocateFlagsInfo flagsInfo = {
```

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_device_funcs.cpp:
```diff
@@ -3796,6 +3827,11 @@ bool WrappedVulkan::Serialise_vkCreateDevice(SerialiserType &ser, VkPhysicalDevi
+    // If BDA is not available, then neither is RT
+    if(!Extensions.contains(VK_EXT_BUFFER_DEVICE_ADDRESS_EXTENSION_NAME) &&
+       !Extensions.contains(VK_KHR_BUFFER_DEVICE_ADDRESS_EXTENSION_NAME))
+      Extensions.removeOne(VK_KHR_ACCELERATION_STRUCTURE_EXTENSION_NAME);
```
This is not the right point to enforce this requirement I think - at this point we are about to create the device on replay, if we remove KHR_acceleration_structure we just make the replay even more invalid by removing the extension that is about to be used.

I think either we should treat this as we do most things with RenderDoc - expect valid API use (which requires the dependency extension KHR_bda to be enabled by the application) and assume that unconditionally, or else this constraint should be enforced at _capture_ time by failing the device create if the dependency extension isn't present.

Note that EXT_bda is not a dependency of KHR_acceleration_structure and can't be used in that way.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_resource_funcs.cpp:
```diff
@@ -3197,13 +3204,27 @@ VkResult WrappedVulkan::vkCreateAccelerationStructureKHR(
+      serialisedCreateInfo.createFlags |=
+          VK_ACCELERATION_STRUCTURE_CREATE_DEVICE_ADDRESS_CAPTURE_REPLAY_BIT_KHR;
+
+      const VkAccelerationStructureDeviceAddressInfoKHR getInfo{
```
Please use normal variable initialisation here too.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_resource_funcs.cpp:
```diff
@@ -3223,6 +3244,13 @@ VkResult WrappedVulkan::vkCreateAccelerationStructureKHR(
       record->memSize = pCreateInfo->size;
+
+      GetResourceManager()->MarkDirtyResource(id);
+      if(pCreateInfo->type == VK_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL_KHR ||
```
I think it would be good to leave a comment here saying that we force-reference bottom levels because it's not feasible to track which top-levels they're referenced by, and that generics could be a bottom level or a top level so we treat them as bottom levels conservatively.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.h:
```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <unordered_map>
+#include <unordered_set>
```
These includes are unnecessary now.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.h:
```diff
@@ -0,0 +1,61 @@
+                 CaptureState state);
+
+  // Called when the initial state is applied.  The AS data is copied from the upload buffer into
+  // the replay buffer
```
nit: the data is copied into an AS, not a buffer.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,352 @@
+constexpr VkDeviceSize asBufferAlignment = 256;
+}
+
+bool VulkanAccelerationStructureManager::Prepare(VkAccelerationStructureKHR as,
```
From experience I think it would be a good idea to explicitly name this `unwrappedAs`. We don't have type information to help us distinguish wrapped from unwrapped objects because of how we handle function hooking and wrapping, so naming conventions are very helpful.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,352 @@
+  VkMemoryRequirements mrq = {};
+  ObjDisp(d)->GetBufferMemoryRequirements(Unwrap(d), dstBuf, &mrq);
+
+  mrq.alignment = asBufferAlignment;
```
I doubt buffer alignment would ever be worse than 256 bytes but this should be `RDCMAX(mrq.alignment, asBufferAlignment)` just to be safe.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,352 @@
+    const VkBufferCreateInfo bufInfo = {VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO, NULL, 0, contentsSize,
+                                        VK_BUFFER_USAGE_TRANSFER_SRC_BIT |
+                                            VK_BUFFER_USAGE_TRANSFER_DST_BIT |
+                                            VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT};
```
Minor nit: if you add a trailing `,` to larger struct definitions like this then clang-format will do a better job. No trailing `,` is usually most useful for struct/array definitions that are small enough to fit in one line or so.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,352 @@
+    VkMemoryRequirements mrq = {};
+    m_pDriver->vkGetBufferMemoryRequirements(d, uploadBuf, &mrq);
+
+    mrq.alignment = asBufferAlignment;
```
Again this should be max'd.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -0,0 +1,352 @@
+                                              Unwrap(initial.buf)};
+  const VkDeviceAddress uploadBufAddr = ObjDisp(d)->GetBufferDeviceAddressKHR(Unwrap(d), &addrInfo);
+
+  VkCopyMemoryToAccelerationStructureInfoKHR copyInfo{
```
Another variable initialisation that should be in normal format.

baldurk reviewed (changes requested):
This looks much better now, thank you for making those changes!

I've left a few notes but they are all stylistic/minor things. Once they're sorted I won't merge this just yet as I'll wait until after the next version which is due out shortly.

cmannett85-arm commented:
Added the broken Mali workaround, and now working fine on NV too.

cmannett85-arm committed:
Vulkan acceleration structure manager

Manages the capturing, serialising, and replay of acceleration structures for Vulkan.  It works in a similar way to how device memory is handled:
* A temporary host-accessible buffer is created for each AS
* The AS is serialised (in the Vulkan AS sense of the word) into it
* The buffer handle is stored in the initial contents and downloaded when appropriate
* Replay is handled similarly but in reverse
* Workaround added for broken Mali AS serialising
Note this is missing the descriptor and SPIR-V handling, that work will follow in a later PR.

The serialiser version has been bumped for backwards compatibility with 'RenderDoc for Arm GPUs 2024.0'.

baldurk commented:
Merged manually to fix the vcxproj change.
